rust/src/dumpfile.rs

```diff
@@ -0,0 +1,391 @@
+            .ok_or_else(|| anyhow!("Missing . in mtime"))?;
+        Ok(Self {
+            sec: u64::from_str(sec)?,
+            nsec: u64::from_str(nsec)?,
+        })
+    }
+}
+
+impl<'k> Xattr<'k> {
+    fn parse(s: &'k str) -> Result<Self> {
+        let (key, value) = s
+            .split_once('=')
+            .ok_or_else(|| anyhow!("Missing = in xattrs"))?;
+        let key = unescape_to_osstr(key)?;
+        let value = unescape(value)?;
+        Ok(Self { key, value })
+    }
+}
+
+impl<'p> Entry<'p> {
+    /// Parse an entry from a composefs dump file line.
+    pub fn parse(s: &'p str) -> Result<Entry<'p>> {
+        let mut components = s.split(' ');
+        let mut next = |name: &str| components.next().ok_or_else(|| anyhow!("Missing {name}"));
+        let path = unescape_to_path(next("path")?)?;
+        let size = u64::from_str(next("size")?)?;
+        let modeval = next("mode")?;
+        let (is_hardlink, mode) = if let Some((_, rest)) = modeval.split_once('@') {
+            (true, u32::from_str_radix(rest, 8)?)
+        } else {
+            (false, u32::from_str_radix(modeval, 8)?)
+        };
+        let nlink = u32::from_str(next("nlink")?)?;
+        let uid = u32::from_str(next("uid")?)?;
+        let gid = u32::from_str(next("gid")?)?;
+        let rdev = u32::from_str(next("rdev")?)?;
+        let mtime = Mtime::from_str(next("mtime")?)?;
+        let payload = optional_str(next("payload")?);
+        let content = optional_str(next("content")?);
+        let fsverity_digest = optional_str(next("digest")?);
+        let xattrs = components.map(Xattr::parse).collect::<Result<Vec<_>>>()?;
+
+        let item = if is_hardlink {
+            let target = unescape_to_path(payload.ok_or_else(|| anyhow!("Missing payload"))?)?;
+            Item::Hardlink { target }
+        } else {
+            match libc::S_IFMT & mode {
+                libc::S_IFREG => Item::Regular {
+                    size,
+                    nlink,
+                    inline_content: content.map(unescape).transpose()?,
+                    fsverity_digest: fsverity_digest.map(ToOwned::to_owned),
+                },
+                libc::S_IFLNK => {
+                    let target =
+                        unescape_to_path(payload.ok_or_else(|| anyhow!("Missing payload"))?)?;
+                    Item::Symlink { nlink, target }
+                }
+                libc::S_IFIFO => Item::Fifo { nlink },
+                libc::S_IFCHR | libc::S_IFBLK => Item::Device { nlink, rdev },
+                libc::S_IFDIR => Item::Directory { size, nlink },
+                o => {
+                    anyhow::bail!("Unhandled mode {o:o}")
+                }
+            }
+        };
+        Ok(Entry {
+            path,
+            uid,
+            gid,
+            mode,
+            mtime,
+            item,
+            xattrs,
+        })
+    }
+
+    /// Remove internal entries
+    /// FIXME: This is arguably a composefs-info dump bug?
```

Comment:
cc @alexlarsson right?  I think these xattrs are internal implementation details and should not leak back into `composefs-info dump` right?

Reply:
Yes, and we should fix this.